### PR TITLE
feat: SSAI tracking

### PR DESCRIPTION
**Pull request by karpov-kir**
Related to: https://bitmovin.atlassian.net/browse/TA-2617.

- Extract tracking logic from `ConvivaAnalytics` into `ConvivaAnalyticsTracker`
- Wire `ConvivaAnalyticsTracker` to player events in `ConvivaAnalytics`
- Introduce `ConvivaAnalyticsSsai` and use `ConvivaAnalyticsTracker` inside to report server-side ad
- Add simulation of server-side ad to the example
- Add tests

<details>
<summary>Examples</summary>

### Server-side ad

image

### Client-side ad

image

### Normal tracking

image

</details>

### CODE REVIEW

File: src/ts/helper/AdHelper.ts [hunk 1]
```diff
@@ -1,8 +1,47 @@
 import { Ad, AdBreak, AdBreakEvent, AdData, AdEvent, ErrorEvent, LinearAd, PlayerAPI, VastAdData } from 'bitmovin-player';
 import * as Conviva from '@convivainc/conviva-js-coresdk';
+import { INTEGRATION_VERSION_CONTENT_METADATA_CUSTOM_TAG } from '../ConvivaAnalyticsTracker';
+
+export interface ServerSideAdInfo {
```
Comment (strangesource) [inline src/ts/helper/AdHelper.ts]: NIT: this is called `SsaiAdInfo` on iOS, maybe we can align here.

File: src/ts/helper/AdHelper.ts [hunk 2]
```diff
@@ -78,29 +107,75 @@ export class AdHelper {
...
-      adInfo[Conviva.Constants.STREAM_URL] = ad.mediaFileUrl;
-    }
+    return adInfo;
+  }
 
-    if ('duration' in ad && ad.duration) {
-      adInfo[Conviva.Constants.DURATION] = ad.duration;
+  public static convertServerSideAdInfoToConvivaAdInfo(serverSideAdInfo: ServerSideAdInfo, allCurrentContentMetadata: Conviva.ConvivaMetadata): Conviva.ConvivaMetadata {
+    const keysToPick = [
+      INTEGRATION_VERSION_CONTENT_METADATA_CUSTOM_TAG,
```
Comment (strangesource) [inline src/ts/helper/AdHelper.ts:142]: On mobile we also copy the custom `streamType` tag.

**Review by strangesource — COMMENTED** [review]
Had a look without deeper knowledge in the codebase, looks good to me. 👍 
Just some minor remarks.

File: src/ts/helper/AdHelper.ts (hunk 2 above)
Comment (karpov-kir) [reply src/ts/helper/AdHelper.ts:142]: There is no such tag on web explicitly:

image

But `IS_LIVE` is resolved to `streamType`:

image

I checked it back then and verified that it worked as expected. `IS_LIVE` is resolved to `Conviva.streamType` and from it Conviva infers the actual is life flag. 😄

File: src/ts/helper/AdHelper.ts (hunk 1 above)
Comment (karpov-kir) [reply src/ts/helper/AdHelper.ts]: Thank you for looking into PR! I've renamed to `SsaiAdInfo` - https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/112/commits/8a5eb0e499765f27d78a3051b0dd42ca7a6a4e00.

File: src/ts/helper/AdHelper.ts (hunk 2 above)
Comment (strangesource) [reply src/ts/helper/AdHelper.ts:142]: The name is a bit ambiguous, it's also used as metadata [here](https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/blob/e607cc8fab8cfea4bb2b88cf39945c3859c33326/src/ts/ConvivaAnalytics.ts#L544) where it contains dash, hls or progressive if I am not mistaken.

File: src/ts/helper/AdHelper.ts [hunk 3]
```diff
@@ -1,8 +1,47 @@
...
+   * A list of ad metadata can be found here: <a href="https://pulse.conviva.com/learning-center/content/sensor_developer_center/sensor_integration/javascript/javascript_stream_sensor.htm#IntegrateAdManagers">Conviva documentation</a>
+   * <p>
+   * Metadata provided here will supersede any data provided in the ad break info.
+   */
+  additionalMetadata?: Record<string, string>;
+}
 
 export class AdHelper {
-  public static mapAdPosition(
+  public static mapClientSideAdPosition(
```
Comment (dweinber) [inline src/ts/helper/AdHelper.ts]: Currently we use sometimes `...ClientSideAd...` namings for methods, and other times `...Csai...`. I'd prefer a more unified naming (not just this occasion here but I'm only adding one comment).

File: src/ts/helper/AdHelper.ts [hunk 4]
```diff
@@ -78,29 +107,75 @@ export class AdHelper {
...
+
+      // TODO This is not exposed currently. Add it whenever the player
+      // exposes it. Related discussion https://bitmovin.slack.com/archives/C0LJ16JBS/p1716801970037469.
+      'c3.ad.mediaFileApiFramework': 'NA',
+
+      // TODO these two are not exposed currently. Add them whenever the player
+      // exposes them similar to https://github.com/bitmovin-engineering/player-android/pull/3147.
+      // Related discussion https://bitmovin.slack.com/archives/C0LJ16JBS/p1716801796326889.
+      'c3.ad.firstAdSystem': 'NA',
+      'c3.ad.firstCreativeId': 'NA',
```
Comment (dweinber) [inline src/ts/helper/AdHelper.ts:125]: Not sure if we really add internal links to the public repo. Usually we try to avoid that.

File: src/ts/ConvivaAnalytics.ts [hunk 5]
```diff
@@ -3,1140 +3,281 @@ import type {
...
-      : Conviva.Constants.LogLevel.NONE;
-
-    Conviva.Analytics.init(customerKey, callbackFunctions, settings);
-
-    this.contentMetadataBuilder = new ContentMetadataBuilder(this.logger);
-
-    this.registerPlayerEvents();
   }
 
-  /**
```
Comment (dweinber) [inline src/ts/ConvivaAnalytics.ts:232]: Why did you remove the documentation for all the public methods? 🤔

File: src/ts/ConvivaAnalyticsTracker.ts [hunk 6]
```diff
@@ -0,0 +1,939 @@
...
+  private readonly logger: Conviva.LoggingInterface;
+  private sessionKey: number;
+  private convivaVideoAnalytics: Conviva.VideoAnalytics;
+  private convivaAdAnalytics: Conviva.AdAnalytics;
+
+  /**
+   * Tracks the ad break status and is true between ON_AD_STARTED and ON_AD_FINISHED/SKIPPED/ERROR.
+   * This flag is required because player.isAd() is unreliable and not always true between the events.
+   */
+  private _isAdBreakActive: boolean;
```
Comment (dweinber) [inline src/ts/ConvivaAnalyticsTracker.ts:131]: What about using [player.ads.getActiveAdBreak()](https://cdn.bitmovin.com/player/web/8/docs/interfaces/Advertising.PlayerAdvertisingAPI.html#getActiveAdBreak) instead, is that reliable? Or would that not work now anyway due to SSAI?

File: src/ts/ConvivaAnalyticsTracker.ts [hunk 7]
```diff
@@ -0,0 +1,939 @@
...
+  /**
+   * Initializes a new conviva tracking session.
+   *
+   * Warning: The integration can only be validated without external session managing. So when using this method we can
+   * no longer ensure that the session is managed at the correct time. Additional: Since some metadata attributes
+   * relies on the players source we can't ensure that all metadata attributes are present at session creation.
+   * Therefore it could be that there will be a 'ContentMetadata created late' issue after conviva validation.
+   *
+   * If no source was loaded and no assetName was set via updateContentMetadata this method will throw an error.
+   */
```
Comment (dweinber) [inline src/ts/ConvivaAnalyticsTracker.ts]: I think all these documentation blocks for methods actually belong to `ConvivaAnalytics.ts` as this is for the user of the integration.

File: spec/tests/PlayerEvents.spec.ts [hunk 8]
```diff
@@ -196,13 +196,15 @@ describe('player event tests', () => {
...
             Conviva.Constants.Playback.SEEK_STARTED,
+            expect.any(Number)
           );
         });
 
         it('on timeshift', () => {
           playerEventHelper.fireTimeShiftEvent();
           expect(MockHelper.latestVideoAnalytics.reportPlaybackMetric).toHaveBeenCalledWith(
             Conviva.Constants.Playback.SEEK_STARTED,
+            -1
```
Comment (dweinber) [inline spec/tests/PlayerEvents.spec.ts:207]: Why `-1`? 🤔

**Review by dweinber — CHANGES_REQUESTED** [review]
Thanks for breaking up the large `ConvivaAnalytics` file! Made reviewing of the feature quite a bit harder, should have probably been a separate PR 😅 

There are a few remarks, please address/clarify, thanks.

**Comment by karpov-kir** [comment]
@dweinber, 

> Thanks for breaking up the large `ConvivaAnalytics` file! Made reviewing of the feature quite a bit harder, should have probably been a separate PR 😅 

Yeah, I thought about it, but the whole feature is almost in this one file [src/ts/ConvivaAnalyticsSsai.ts](https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/112/files#diff-42695249a7c5f7001137a6dc9531ed28be3327f5b78901c53bfd9ae1a71afc9a). 🙂

File: src/ts/helper/AdHelper.ts (hunk 3 above)
Comment (karpov-kir) [reply src/ts/helper/AdHelper.ts]: Unified in https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/112/commits/0045c6f9521e1124369a60a9193b45eb2534cfe0.

File: src/ts/helper/AdHelper.ts (hunk 4 above)
Comment (karpov-kir) [reply src/ts/helper/AdHelper.ts:125]: These links were in the repo already from the previous PR - https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/104#discussion_r1625652063.

File: src/ts/ConvivaAnalytics.ts (hunk 5 above)
Comment (karpov-kir) [reply src/ts/ConvivaAnalytics.ts:232]: Because they are now inside `ConvivaAnalyticsTracker`. But it looks like this way they are not exposed to users. So, I'll move them back.

File: src/ts/ConvivaAnalyticsTracker.ts (hunk 6 above)
Comment (karpov-kir) [reply src/ts/ConvivaAnalyticsTracker.ts:131]: `isAdBreak` was in the previous implementation, I simply followed it, so not sure if [player.ads.getActiveAdBreak()](https://cdn.bitmovin.com/player/web/8/docs/interfaces/Advertising.PlayerAdvertisingAPI.html#getActiveAdBreak) is reliable. Anyways, we need this flag because it also tacks SSAI, so we cannot use `getActiveAdBreak`.

File: spec/tests/PlayerEvents.spec.ts (hunk 8 above)
Comment (karpov-kir) [reply spec/tests/PlayerEvents.spec.ts:207]: It was this way always https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/blob/v5.0.0/src/ts/ConvivaAnalytics.ts#L759:

image

But there was a problem with passing this value around in the previous implementation, so it's fixed in this PR.

File: src/ts/ConvivaAnalytics.ts (hunk 5 above)
Comment (karpov-kir) [reply src/ts/ConvivaAnalytics.ts:232]: Moved in https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/112/commits/5803fb7253bcc383ceaec4f74a6d89255e1f09e1.

File: src/ts/ConvivaAnalyticsTracker.ts (hunk 7 above)
Comment (karpov-kir) [reply src/ts/ConvivaAnalyticsTracker.ts]: Moved in https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/112/commits/5803fb7253bcc383ceaec4f74a6d89255e1f09e1.

File: src/ts/helper/AdHelper.ts (hunk 2 above)
Comment (karpov-kir) [reply src/ts/helper/AdHelper.ts:142]: Aha, I see https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/112/commits/65874d18c824c9d93cb963820bf68771f84ce978.

**Review by strangesource — APPROVED** [review]

File: src/ts/helper/AdHelper.ts (hunk 4 above)
Comment (dweinber) [reply src/ts/helper/AdHelper.ts:125]: Okay 🙈

File: src/ts/ConvivaAnalyticsTracker.ts (hunk 6 above)
Comment (dweinber) [reply src/ts/ConvivaAnalyticsTracker.ts:131]: Makes sense regarding SSAI 👍

File: spec/tests/PlayerEvents.spec.ts (hunk 8 above)
Comment (dweinber) [reply spec/tests/PlayerEvents.spec.ts:207]: Thanks for clarifying!

**Review by dweinber — APPROVED** [review]
🎉
